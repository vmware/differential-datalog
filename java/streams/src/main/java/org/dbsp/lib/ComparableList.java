/*
 * Copyright (c) 2021 VMware, Inc.
 * SPDX-License-Identifier: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.dbsp.lib;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * Lists that are compared elementwise.
 */
public class ComparableList<T extends Comparable<T>>
        extends ArrayList<T>
        implements Comparable<ComparableList<T>> {
    public ComparableList(int capacity) {
        super(capacity);
    }

    @SafeVarargs
    public ComparableList(T... o) {
        super(o.length);
        this.addAll(Arrays.asList(o));
    }

    @Override
    public int compareTo(ComparableList<T> o) {
        for (int i = 0; i < this.size() && i < o.size(); i++) {
            int compare = this.get(i).compareTo(o.get(i));
            if (compare != 0)
                return compare;
        }
        return Integer.compare(this.size(), o.size());
    }
}
