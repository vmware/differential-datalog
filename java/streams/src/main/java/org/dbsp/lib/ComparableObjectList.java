/*
 * Copyright (c) 2021 VMware, Inc.
 * SPDX-License-Identifier: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.dbsp.lib;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * A dynamically typed version of ComparableList.
 */
public class ComparableObjectList
    extends ArrayList<Object>
        implements Comparable<ComparableObjectList>
        // implements ComparableObject
{
    public ComparableObjectList(Object... o) {
        super(o.length);
        this.addAll(Arrays.asList(o));
    }

    @SuppressWarnings("unchecked")
    @Override
    public int compareTo(ComparableObjectList o) {
        for (int i = 0; i < this.size() && i < o.size(); i++) {
            // The following may not implement ComparableObject, but they may actually have a compareTo method.
            // We cannot retrofit this interface to existing classes like Integer or String.
            int compare = ((Comparable<Object>)this.get(i)).compareTo((Comparable<Object>)o.get(i));
            if (compare != 0)
                return compare;
        }
        return Integer.compare(this.size(), o.size());
    }
}
