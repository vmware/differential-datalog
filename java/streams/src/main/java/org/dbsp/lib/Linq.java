/*
 * Copyright (c) 2021 VMware, Inc.
 * SPDX-License-Identifier: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package org.dbsp.lib;

import org.dbsp.circuits.types.Type;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;

/**
 * Some utility classes inspired by C# Linq.
 */
public class Linq {
    public static void indent(int amount, StringBuilder builder) {
        for (int i = 0; i < amount; i++)
            builder.append(" ");
    }

    static class MapIterator<T, S> implements Iterator<S> {
        final Iterator<T> data;
        final Function<T, S> map;

        MapIterator(Iterator<T> data, Function<T, S> function) {
            this.data = data;
            this.map = function;
        }

        @Override
        public boolean hasNext() {
            return this.data.hasNext();
        }

        public S next() {
            T next = this.data.next();
            return this.map.apply(next);
        }
    }

    static class MapIterable<T, S> implements Iterable<S> {
        final MapIterator<T, S> mapIterator;

        MapIterable(Iterable<T> data, Function<T, S> function) {
            this.mapIterator = new MapIterator<T, S>(data.iterator(), function);
        }

        @Override
        public Iterator<S> iterator() {
            return this.mapIterator;
        }
    }

    public static <T, S> Iterable<S> map(Iterable<T> data, Function<T, S> function) {
        return new MapIterable<T, S>(data, function);
    }

    public static <T, S> Iterator<S> map(Iterator<T> data, Function<T, S> function) {
        return new MapIterator<T, S>(data, function);
    }

    public static <T, S> List<S> map(List<T> data, Function<T, S> function) {
        List<S> result = new ArrayList<S>(data.size());
        for (T aData : data)
            result.add(function.apply(aData));
        return result;
    }

    public static <T, S> List<S> flatMap(List<T> data, Function<T, List<S>> function) {
        List<S> result = new ArrayList<S>(data.size());
        for (T aData : data)
            result.addAll(function.apply(aData));
        return result;
    }

    public static <T, S> List<S> as(List<T> data, Class<S> sc) {
        return Linq.map(data, sc::cast);
    }

    public static <T, S> S[] map(T[] data, Function<T, S> function, Class<S> sc) {
        @SuppressWarnings("unchecked")
        S[] result = (S[])Array.newInstance(sc, data.length);
        for (int i=0; i < data.length; i++)
            result[i] = function.apply(data[i]);
        return result;
    }

    @SafeVarargs
    public static <T> List<T> list(T... data) {
        return Arrays.asList(data);
    }

    public static <T> List<T> where(List<T> data, Predicate<T> function) {
        List<T> result = new ArrayList<T>();
        for (T aData : data)
            if (function.test(aData))
                result.add(aData);
        return result;
    }

    @SuppressWarnings("unchecked")
    public static <T> T[] where(T[] data, Predicate<T> function) {
        List<T> result = new ArrayList<T>();
        for (T datum : data)
            if (function.test(datum))
                result.add(datum);
        return (T[]) result.toArray();
    }

    public static <T> boolean any(Iterable<T> data, Predicate<T> test) {
        for (T d: data)
            if (test.test(d)) {
                return true;
            }
        return false;
    }
}
