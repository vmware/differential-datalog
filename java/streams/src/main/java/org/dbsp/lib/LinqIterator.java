/*
 * Copyright (c) 2021 VMware, Inc.
 * SPDX-License-Identifier: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.dbsp.lib;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;

/**
 * An iterator that returns values of type T.
 * These iterators cannot be reset.
 * @param <T>  Value returned.
 */
public abstract class LinqIterator<T> implements Iterator<T> {
    /**
     * An iterator over values of type S.
     * @param map  Function that converts each value of type T to a value of type S.
     * @param <S>  Value produced by function.
     */
    public <S> LinqIterator<S> map(Function<T, S> map) {
        return new LinqIterator<S>() {
            @Override
            public boolean hasNext() {
                return LinqIterator.this.hasNext();
            }

            @Nullable
            @Override
            public S next() {
                return map.apply(LinqIterator.this.next());
            }
        };
    }

    public LinqIterator<T> filter(Predicate<T> predicate) {
        return new LinqIterator<T>() {
            @Nullable
            T next = null;
            // True if we are done iterating over source.
            boolean done = false;
            // True if we don't know whether there are more elements.
            boolean needScan = true;

            /**
             * Sets 'next' to the next value that passes the predicate.
             * Sets 'done' to 'true' if the source iterator is not done.
             */
            void findNext() {
                if (!needScan)
                    return;
                this.needScan = false;
                do {
                    if (!LinqIterator.this.hasNext()) {
                        this.done = true;
                        return;
                    }
                    this.next = LinqIterator.this.next();
                } while (!predicate.test(this.next));
            }

            @Override
            public boolean hasNext() {
                this.findNext();
                return !this.done;
            }

            @Nullable
            @Override
            public T next() {
                this.findNext();
                if (this.done)
                    throw new RuntimeException("No more values");
                this.needScan = true;
                return this.next;
            }
        };
    }

    public <S> LinqIterator<Pair<T, S>> zip(LinqIterator<S> other) {
        return new LinqIterator<Pair<T, S>>() {
            @Override
            public boolean hasNext() {
                return LinqIterator.this.hasNext() && other.hasNext();
            }

            @Override
            public Pair<T, S> next() {
                return new Pair<T, S>(LinqIterator.this.next(), other.next());
            }
        };
    }

    public <S, U> LinqIterator<Triple<T, S, U>> zip3(LinqIterator<S> left, LinqIterator<U> right) {
        return new LinqIterator<Triple<T, S, U>>() {
            @Override
            public boolean hasNext() {
                return LinqIterator.this.hasNext() && left.hasNext() && right.hasNext();
            }

            @Override
            public Triple<T, S, U> next() {
                return new Triple<T, S, U>(LinqIterator.this.next(), left.next(), right.next());
            }
        };
    }

    /**
     * Zip two itrators; returns an itrator of pairs with elements from both iterators.
     * Stops when the first iterator terminates.
     * @param other  List to zip with.
     * @param <S>    Type of elements in the other list.
     * @return       A list with the length the shorter of this and other.
     */
    public <S> LinqIterator<Pair<T, S>> zip(List<S> other) {
        return this.zip(LinqIterator.create(other));
    }

    public <S, U> LinqIterator<Triple<T, S, U>> zip3(List<S> left, List<U> right) {
        return this.zip3(LinqIterator.create(left), LinqIterator.create(right));
    }

    public List<T> toList() {
        ArrayList<T> result = new ArrayList<T>();
        while (this.hasNext()) {
            result.add(this.next());
        }
        return result;
    }

    public static <T> LinqIterator<T> create(List<T> list) {
        return create(list.iterator());
    }

    public static <T> LinqIterator<T> create(T[] data) {
        return new LinqIterator<T>() {
            int index = 0;

            @Override
            public boolean hasNext() {
                return this.index < data.length;
            }

            @Override
            public T next() {
                return data[this.index++];
            }
        };
    }

    public static <T> LinqIterator<T> create(Iterator<T> data) {
        return new LinqIterator<T>() {
            @Override
            public boolean hasNext() {
                return data.hasNext();
            }

            @Override
            public T next() {
                return data.next();
            }
        };
    }

    public boolean any(Predicate<T> p) {
        return this.filter(p).hasNext();
    }
}
