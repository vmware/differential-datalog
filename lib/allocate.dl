/* Allocate a unique identifier in the range `min_val..max_val` for each element in `toallocate`.
 * `allocated` stores already allocated ids that must not be re-used.
 *
 * Returns new allocations.  May only assign a subset of values if the address space has been
 * exhausted.
 *
 * `N` must be a numeric type, i.e., its Rust representation must have `Num`, `Add`, `Ord`, and
 * `Copy` traits.  However, as DDlog does not have traits, it will happily compile programs with
 * any `N` type, leaving it to the Rust compiler to detect the unsatisfied trait.
 */
extern function allocate(
    allocated: Set<'N>,
    toallocate: Vec<'B>,
    min_val: 'N,
    max_val: 'N): Vec<('B, 'N)>

/* Allocate a unique identifier in the range `min_val..max_val` for each element in `toallocate`;
 * use optional hint as starting point to search for available identifier.
 * `allocated` stores already allocated ids that must not be re-used.
 *
 * Returns new allocations.  May only assign a subset of values if the address space has been
 * exhausted.
 *
 * `N` must be a numeric type, i.e., its Rust representation must have `Num`, `Add`, `Ord`, and
 * `Copy` traits.  However, as DDlog does not have traits, it will happily compile programs with
 * any `N` type, leaving it to the Rust compiler to detect the unsatisfied trait.
 */
extern function allocate_with_hint(
    allocated: Set<'N>,
    toallocate: Vec<('B, Option<'N>)>,
    min_val: 'N,
    max_val: 'N): Vec<('B, 'N)>

/* Allocate a unique identifier in the range `min_val..max_val` for each element in `toallocate`.
 * `allocated` stores already allocated ids that must not be re-used.
 *
 * Returns new allocations.  After exhausting the address space, assigns `None` to all remaining
 * items.
 *
 * `N` must be a numeric type, i.e., its Rust representation must have `Num`, `Add`, `Ord`, and
 * `Copy` traits.  However, as DDlog does not have traits, it will happily compile programs with
 * any `N` type, leaving it to the Rust compiler to detect the unsatisfied trait.
 */
extern function allocate_opt(
    allocated: Set<'N>,
    toallocate: Vec<'B>,
    min_val: 'N,
    max_val: 'N): Vec<('B, Option<'N>)>

/* Allocate a unique identifier in the range `min_val..max_val` for each element in `toallocate`,
 * while preserving existing allocations in `allocated`.
 *
 * For each value in `toallocate`, either returns existing allocation in `allocated` or allocates
 * an unused value in the range.
 *
 * `N` must be a numeric type, i.e., its Rust representation must have `Num`, `Add`, `Ord`, and
 * `Copy` traits.  However, as DDlog does not have traits, it will happily compile programs with
 * any `N` type, leaving it to the Rust compiler to detect the unsatisfied trait.
 */
extern function adjust_allocation(
    allocated: Map<'A, 'N>,
    toallocate: Vec<'A>,
    min_val: 'N,
    max_val: 'N): Vec<('A, 'N)>
