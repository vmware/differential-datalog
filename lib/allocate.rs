extern crate num;
use std::collections::BTreeSet;
use std::ops;
use std::cmp;

pub fn allocate_allocate<B: Ord+Clone, N:num::Num+ops::Add+cmp::Ord+Copy>(
    allocated: &std_Set<N>,
    toallocate: &std_Vec<B>,
    min_val: &N,
    max_val: &N) -> std_Vec<(B,N)>
{
    assert!(*max_val >= *min_val);
    let range = *max_val - *min_val;

    // Next index to consider
    let mut next = *allocated.x.range(..).next_back().unwrap_or(&max_val);
    // allocated may contain values outside of the [min_val..max_val] range.
    if next < *min_val || next > *max_val {
        next = *max_val;
    };
    let mut offset = N::zero();
    let mut res = std_Vec::new();
    for b in toallocate.x.iter() {
        loop {
            next = if next == *max_val { *min_val } else { next + N::one() };
            offset = offset + N::one();

            if allocated.x.contains(&next) {
                if offset == range + N::one() { return res; };
                continue;
            } else {
                res.x.push(((*b).clone(), next));
                if offset == range + N::one() { return res; };
                break;
            }
        }
    };
    return res
}

pub fn allocate_adjust_allocation<A: Ord+Clone, N: num::Num+ops::Add+cmp::Ord+Copy>(
    allocated: &std_Map<A, N>,
    toallocate: &std_Vec<A>,
    min_val: &N,
    max_val: &N) -> std_Vec<(A,N)>
{
    assert!(*max_val >= *min_val);
    let range = *max_val - *min_val;

    let allocated_ids: BTreeSet<N> = allocated.x.values().cloned().collect();
    let mut next = *allocated_ids.range(..).next_back().unwrap_or(&max_val);
    // allocated may contain values outside of the [min_val..max_val] range.
    if next < *min_val || next > *max_val {
        next = *max_val;
    };
    let mut offset = N::zero();
    let mut res = std_Vec::new();
    // Signal that address space has been exhausted, but iteration must continue to
    // preserve existing allocations.
    let mut exhausted = false;
    for b in toallocate.x.iter() {
        match allocated.x.get(b) {
            Some(x) => { res.push(((*b).clone(), x.clone())); },
            None => loop {
                if exhausted { break; };
                if offset == range { exhausted = true; };
                next = if next == *max_val { *min_val } else { next + N::one() };
                offset = offset + N::one();

                if allocated_ids.contains(&next) {
                    continue;
                } else {
                    res.x.push(((*b).clone(), next));
                    break;
                }
            }
        }
    };
    return res
}
