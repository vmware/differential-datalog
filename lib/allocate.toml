[dependencies.num]
version = "0.2"
