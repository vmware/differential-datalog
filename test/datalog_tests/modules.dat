start;

insert Rmain(foolib.lib.Tlib{0}, foolib.ns1.m1.T1{1}),
insert foolib.lib.Rlib(foolib.ns1.m2.T2{2}),
insert foolib.ns1.m1.R1(foolib.ns1.m2.T2{3}),
insert foolib.ns1.m2.R2(true);

commit;

dump;
