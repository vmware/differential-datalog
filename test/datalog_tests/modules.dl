import foolib.lib
import foolib.ns1.m1 as m1
import foolib.ns1.m2 as m2

input relation Rmain(f1: Tlib, f2: m1.T1)
output relation Rout(f1: bool, f2: bigint)

Rout(q, x) :- Rmain(Tlib{x}, m1.T1{y}), Rlib(m2.T2{z}), m1.R1(m2.T2{p}), m2.R2(q).

typedef Tmain = Tmain {f1: m1.T1}

function fmain(a1: Rlib): m1.R1 = m1.R1{m2.T2{0}}
