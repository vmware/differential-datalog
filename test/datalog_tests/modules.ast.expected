typedef Rmain = Rmain{f1: foolib.lib.Tlib, f2: foolib.ns1.m1.T1}
typedef Rout = Rout{f1: bool, f2: bigint}
typedef Tmain = Tmain{f1: foolib.ns1.m1.T1}
typedef foolib.lib.Rlib = foolib.lib.Rlib{f1: foolib.ns1.m2.T2}
typedef foolib.lib.Tlib = foolib.lib.Tlib{f1: bigint}
typedef foolib.ns1.m1.R1 = foolib.ns1.m1.R1{f1: foolib.ns1.m2.T2}
typedef foolib.ns1.m1.T1 = foolib.ns1.m1.T1{f1: bigint}
typedef foolib.ns1.m2.R2 = foolib.ns1.m2.R2{f1: bool}
typedef foolib.ns1.m2.T2 = foolib.ns1.m2.T2{f1: bigint}
typedef log.log_level_t = signed<32>
typedef log.module_t = signed<32>
typedef redist.BiEdge = redist.BiEdge{parent: redist.entid_t, child: redist.entid_t}
typedef redist.DdlogBinding = redist.DdlogBinding{tn: redist.tnid_t, entity: redist.entid_t}
typedef redist.DdlogDependency = redist.DdlogDependency{parent: redist.entid_t, child: redist.entid_t}
typedef redist.DdlogNode = redist.DdlogNode{id: redist.entid_t}
typedef redist.LabeledNode = redist.LabeledNode{node: redist.entid_t, scc: redist.entid_t}
typedef redist.SCCBinding = redist.SCCBinding{scc: redist.entid_t, bindings: std.Ref<tinyset.Set64<redist.tnid_t>>}
typedef redist.SCCEdge = redist.SCCEdge{parent: redist.entid_t, child: redist.entid_t}
typedef redist.SCCSpan = redist.SCCSpan{scc: redist.entid_t, span: std.Ref<tinyset.Set64<redist.tnid_t>>}
typedef redist.Span = redist.Span{entity: redist.entid_t, tns: std.Ref<tinyset.Set64<redist.tnid_t>>}
typedef redist.entid_t = bit<32>
typedef redist.tnid_t = bit<16>
typedef std.Either<'A,'B> = std.Left{l: 'A} | std.Right{r: 'B}
extern type std.Group<'A>
extern type std.Map<'K,'V>
typedef std.Option<'A> = std.Some{x: 'A} | std.None{}
#[size = 8]
extern type std.Ref<'A>
extern type std.Set<'A>
extern type std.Vec<'A>
extern type tinyset.Set64<'X>
function fmain (a1: foolib.lib.Rlib): foolib.ns1.m1.R1 =
    foolib.ns1.m1.R1{.f1=foolib.ns1.m2.T2{.f1=0}}
function foolib.ns1.m1.fM1 (a1: foolib.ns1.m1.T1): foolib.ns1.m1.R1 =
    foolib.ns1.m1.R1{.f1=foolib.ns1.m2.T2{.f1=a1.f1}}
extern function log.log (module: log.module_t, level: log.log_level_t, msg: string): bool
function redist.edge_child (e: redist.BiEdge): redist.entid_t =
    e.child
function redist.edge_parent (e: redist.BiEdge): redist.entid_t =
    e.parent
extern function std.__builtin_2string (x: 'X): string
extern function std.deref (x: std.Ref<'A>): 'A
extern function std.group2map (g: std.Group<('K, 'V)>): std.Map<'K,'V>
extern function std.group2set (g: std.Group<'A>): std.Set<'A>
extern function std.group2vec (g: std.Group<'A>): std.Vec<'A>
extern function std.group_count (g: std.Group<'A>): bit<64>
extern function std.group_first (g: std.Group<'A>): 'A
extern function std.group_max (g: std.Group<'A>): 'A
extern function std.group_min (g: std.Group<'A>): 'A
extern function std.group_nth (g: std.Group<'A>, n: bit<64>): std.Option<'A>
extern function std.group_set_unions (g: std.Group<std.Set<'A>>): std.Set<'A>
extern function std.group_setref_unions (g: std.Group<std.Ref<std.Set<'A>>>): std.Ref<std.Set<'A>>
extern function std.group_sum (g: std.Group<'A>): 'A
function std.group_unzip (g: std.Group<('X, 'Y)>): (std.Vec<'X>, std.Vec<'Y>) =
    ((var xs: std.Vec<'X>) = std.vec_empty();
     ((var ys: std.Vec<'Y>) = std.vec_empty();
      (for (v in g) {
           ((var x, var y) = v;
            (std.vec_push(xs, x);
             std.vec_push(ys, y)))
       };
       (xs, ys))))
extern function std.hash128 (x: 'X): bit<128>
extern function std.hash64 (x: 'X): bit<64>
extern function std.hex (x: 'X): string
extern function std.htonl (x: bit<32>): bit<32>
extern function std.htons (x: bit<16>): bit<16>
function std.is_none (x: std.Option<'A>): bool =
    match (x) {
        std.None{} -> true,
        _ -> false
    }
function std.is_some (x: std.Option<'A>): bool =
    match (x) {
        std.Some{.x=_} -> true,
        _ -> false
    }
extern function std.map_contains_key (m: std.Map<'K,'V>, k: 'K): bool
extern function std.map_empty (): std.Map<'K,'V>
extern function std.map_get (m: std.Map<'K,'V>, k: 'K): std.Option<'V>
extern function std.map_insert (m: mut std.Map<'K,'V>, k: 'K, v: 'V): ()
extern function std.map_insert_imm (m: std.Map<'K,'V>, k: 'K, v: 'V): std.Map<'K,'V>
extern function std.map_is_empty (m: std.Map<'K,'V>): bool
extern function std.map_remove (m: mut std.Map<'K,'V>, k: 'K): ()
extern function std.map_singleton (k: 'K, v: 'V): std.Map<'K,'V>
extern function std.map_union (m1: std.Map<'K,'V>, m2: std.Map<'K,'V>): std.Map<'K,'V>
extern function std.max (x: 'A, y: 'A): 'A
extern function std.min (x: 'A, y: 'A): 'A
extern function std.ntohl (x: bit<32>): bit<32>
extern function std.ntohs (x: bit<16>): bit<16>
extern function std.parse_dec_i64 (s: string): std.Option<signed<64>>
extern function std.parse_dec_u64 (s: string): std.Option<bit<64>>
extern function std.pow32 (base: 'A, exp: bit<32>): 'A
extern function std.range (from: 'A, to: 'A, step: 'A): std.Vec<'A>
extern function std.ref_new (x: 'A): std.Ref<'A>
extern function std.set2vec (s: std.Set<'A>): std.Vec<'A>
extern function std.set_contains (s: std.Set<'X>, v: 'X): bool
extern function std.set_empty (): std.Set<'X>
extern function std.set_insert (s: mut std.Set<'X>, v: 'X): ()
extern function std.set_insert_imm (s: std.Set<'X>, v: 'X): std.Set<'X>
extern function std.set_is_empty (s: std.Set<'X>): bool
extern function std.set_nth (s: std.Set<'X>, n: bit<64>): std.Option<'X>
extern function std.set_singleton (x: 'X): std.Set<'X>
extern function std.set_size (s: std.Set<'X>): bit<64>
extern function std.set_union (s1: std.Set<'X>, s2: std.Set<'X>): std.Set<'X>
extern function std.set_unions (sets: std.Vec<std.Set<'X>>): std.Set<'X>
extern function std.str_to_lower (s: string): string
extern function std.string_contains (s1: string, s2: string): bool
extern function std.string_join (strings: std.Vec<string>, sep: string): string
extern function std.string_len (s: string): bit<64>
extern function std.string_split (s: string, sep: string): std.Vec<string>
extern function std.string_substr (s: string, start: bit<64>, end: bit<64>): string
extern function std.vec2set (s: std.Vec<'A>): std.Set<'A>
extern function std.vec_contains (v: std.Vec<'X>, x: 'X): bool
extern function std.vec_empty (): std.Vec<'A>
extern function std.vec_is_empty (v: std.Vec<'X>): bool
extern function std.vec_len (v: std.Vec<'X>): bit<64>
extern function std.vec_nth (v: std.Vec<'X>, n: bit<64>): std.Option<'X>
extern function std.vec_push (v: mut std.Vec<'X>, x: 'X): ()
extern function std.vec_push_imm (v: std.Vec<'X>, x: 'X): std.Vec<'X>
extern function std.vec_singleton (x: 'X): std.Vec<'X>
extern function tinyset.contains (s: tinyset.Set64<'X>, v: 'X): bool
extern function tinyset.empty (): tinyset.Set64<'X>
extern function tinyset.group2set (g: std.Group<'A>): tinyset.Set64<'A>
extern function tinyset.group_set_unions (g: std.Group<tinyset.Set64<'A>>): tinyset.Set64<'A>
extern function tinyset.group_setref_unions (g: std.Group<std.Ref<tinyset.Set64<'A>>>): std.Ref<tinyset.Set64<'A>>
extern function tinyset.insert (s: mut tinyset.Set64<'X>, v: 'X): ()
extern function tinyset.insert_imm (s: tinyset.Set64<'X>, v: 'X): tinyset.Set64<'X>
extern function tinyset.is_empty (s: tinyset.Set64<'X>): bool
extern function tinyset.nth (s: tinyset.Set64<'X>, n: bit<64>): std.Option<bit<64>>
extern function tinyset.singleton (x: 'X): tinyset.Set64<'X>
extern function tinyset.size (s: tinyset.Set64<'X>): bit<64>
extern function tinyset.union (s1: tinyset.Set64<'X>, s2: tinyset.Set64<'X>): tinyset.Set64<'X>
extern function tinyset.unions (sets: std.Vec<tinyset.Set64<'X>>): tinyset.Set64<'X>
extern transformer graph.ConnectedComponents(Edges: relation['E], from: function(e: 'E): 'N, to: function(e: 'E): 'N) -> (CCLabels: relation[('N, 'N)])
extern transformer graph.ConnectedComponents64(Edges: relation['E], from: function(e: 'E): 'N, to: function(e: 'E): 'N) -> (CCLabels: relation[('N, 'N)])
extern transformer graph.SCC(Edges: relation['E], from: function(e: 'E): 'N, to: function(e: 'E): 'N) -> (SCCLabels: relation[('N, 'N)])
extern transformer graph.UnsafeBidirectionalEdges(Edges: relation['E], from: function(e: 'E): 'N, to: function(e: 'E): 'N) -> (BiEdges: relation[('N, 'N)])
input relation Rmain [Rmain]
output relation Rout [Rout]
input relation foolib.lib.Rlib [foolib.lib.Rlib]
input relation foolib.ns1.m1.R1 [foolib.ns1.m1.R1]
input relation foolib.ns1.m2.R2 [foolib.ns1.m2.R2]
relation redist.BiEdge [redist.BiEdge]
input relation redist.DdlogBinding [redist.DdlogBinding]
input relation redist.DdlogDependency [redist.DdlogDependency]
input relation redist.DdlogNode [redist.DdlogNode]
relation redist.LabeledNode [redist.LabeledNode]
relation redist.SCCBinding [redist.SCCBinding]
relation redist.SCCEdge [redist.SCCEdge]
relation redist.SCCLabel [(redist.entid_t, redist.entid_t)]
relation redist.SCCSpan [redist.SCCSpan]
output relation redist.Span [redist.Span]
Rout(.f1=q, .f2=x) :- Rmain(.f1=foolib.lib.Tlib{.f1=x}, .f2=foolib.ns1.m1.T1{.f1=y}), foolib.lib.Rlib(.f1=foolib.ns1.m2.T2{.f1=z}), foolib.ns1.m1.R1(.f1=foolib.ns1.m2.T2{.f1=p}), foolib.ns1.m2.R2(.f1=q).
redist.BiEdge(.parent=parent, .child=child) :- redist.DdlogDependency(.parent=parent, .child=child), redist.DdlogDependency(.parent=child, .child=parent).
redist.LabeledNode(.node=node, .scc=scc) :- redist.SCCLabel[(node, scc)].
redist.LabeledNode(.node=node, .scc=node) :- redist.DdlogNode(.id=node), not redist.SCCLabel[(node, _)].
redist.SCCEdge(.parent=parentscc, .child=childscc) :- redist.DdlogDependency(.parent=parent, .child=child), redist.LabeledNode(.node=parent, .scc=parentscc), redist.LabeledNode(.node=child, .scc=childscc), (childscc != parentscc).
redist.SCCBinding(.scc=scc, .bindings=std.ref_new(bindings)) :- redist.DdlogBinding(.tn=tn, .entity=entity), redist.LabeledNode(.node=entity, .scc=scc), var bindings = Aggregate((scc), tinyset.group2set(tn)).
redist.SCCSpan(.scc=scc, .span=bindings) :- redist.SCCBinding(.scc=scc, .bindings=bindings).
redist.SCCSpan(.scc=parent, .span=tns) :- redist.SCCEdge(.parent=child, .child=parent), redist.SCCSpan(.scc=child, .span=child_tns), var tns = Aggregate((parent), tinyset.group_setref_unions(child_tns)).
redist.Span(.entity=node, .tns=span) :- redist.SCCSpan(.scc=scc, .span=span), redist.LabeledNode(.node=node, .scc=scc).
apply graph.ConnectedComponents(redist.BiEdge, redist.edge_child, redist.edge_parent) -> (redist.SCCLabel)
